src/safeposix/syscalls/fs_calls.rs

```diff
@@ -2261,107 +2261,198 @@ impl Cage {
+    ///
+    /// To learn more about the syscall, valid mode bits, and error values, see
+    /// [chmod(2)](https://man7.org/linux/man-pages/man2/chmod.2.html)
 
     pub fn _chmod_helper(inodenum: usize, mode: u32) {
+        //getting a mutable reference to an inode struct that corresponds to the file whose mode bits we want to change
         let mut thisinode = FS_METADATA.inodetable.get_mut(&inodenum).unwrap();
+        //log is used to store all the changes made to the filesystem. After the cage is closed, all the collected
+        //changes are serialized and the state of the underlying filsystem is persisted. This allows us to avoid
+        //serializing and persisting filesystem state after every chmod_syscall().
         let mut log = true;
+        //This sanity check of the `mode` argument has already been done in `chmod_syscall()` before this helper funciton is called.
+        //Isn't it redundant to check the same thing here again?
         if mode & (S_IRWXA | (S_FILETYPEFLAGS as u32)) == mode {
+            //We obtain the mode bits that should remain intact by bitwise-and'ing the inode's mode bits with 
+            //the set of bits that can be changed via `chmod_syscall`. The changes are applied by bitwise-or'ing 
+            //the intact mode bits with the changed mode bits.
             match *thisinode {
                 Inode::File(ref mut general_inode) => {
-                    general_inode.mode = (general_inode.mode & !S_IRWXA) | mode
+                    general_inode.mode = (general_inode.mode & !S_IRWXA) | (mode & S_IRWXA);
                 }
                 Inode::CharDev(ref mut dev_inode) => {
-                    dev_inode.mode = (dev_inode.mode & !S_IRWXA) | mode;
+                    dev_inode.mode = (dev_inode.mode & !S_IRWXA) | (mode & S_IRWXA);
                 }
                 Inode::Socket(ref mut sock_inode) => {
-                    sock_inode.mode = (sock_inode.mode & !S_IRWXA) | mode;
+                    sock_inode.mode = (sock_inode.mode & !S_IRWXA) | (mode & S_IRWXA);
+                    //Sockets only exist as long as the cages using them are running. After these cages are closed, no changes
+                    //to sockets' inodes need to be persisted, thus using log is unnecessary.
                     log = false;
                 }
                 Inode::Dir(ref mut dir_inode) => {
-                    dir_inode.mode = (dir_inode.mode & !S_IRWXA) | mode;
+                    dir_inode.mode = (dir_inode.mode & !S_IRWXA) | (mode & S_IRWXA);
                 }
             }
             drop(thisinode);
+            //changes to an inode are saved into the log for all file types except for Sockets
             if log {
                 log_metadata(&FS_METADATA, inodenum)
             };
         }
     }
 
-    //------------------------------------CHMOD SYSCALL------------------------------------
-
     pub fn chmod_syscall(&self, path: &str, mode: u32) -> i32 {
```

Comment:
can we add doc string to this function aswell

Reply:
ideally it should describe this function and link to helper function

Reply:
where exactly should I specify that there is a helper function? Should I put it in the description block right after the general description of the syscall?